File: Excel/646 Challenge.py
import pandas as pd

path = "646 Insert Quarterly Total Line.xlsx"
input = pd.read_excel(path, usecols="A:B", skiprows=1, nrows=12)
test = pd.read_excel(path, usecols="D:E", skiprows=1, nrows=17).rename(columns=lambda x: x.split('.')[0])

def insert_quarterly_totals(df):
    quarters = {'Q1': ['Jan', 'Feb', 'Mar'], 'Q2': ['Apr', 'May', 'Jun'], 
                'Q3': ['Jul', 'Aug', 'Sep'], 'Q4': ['Oct', 'Nov', 'Dec']}
    result = pd.concat([df[df['Data'].isin(months)]._append(
        {'Data': 'Quarter Total', 'Value': df[df['Data'].isin(months)]['Value'].sum()}, ignore_index=True)
        for months in quarters.values()])
    return result.reset_index(drop=True)

input_with_totals = insert_quarterly_totals(input)
print(all(input_with_totals == test)) # True

File: Excel/600-699/646/646 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/646 Insert Quarterly Total Line.xlsx"
input = read_excel(path, range = "A2:B14")
test  = read_excel(path, range = "D2:E18")

result = input %>%
  mutate(Quarter = rep(1:4, each = 3)) 
qt = result %>%
  summarise(Data = "Quarter Total", Value = sum(Value), .by = Quarter) %>%
  bind_rows(result) %>%
  arrange(Quarter, grepl("Total", Data)) %>%
  select(-Quarter)

all.equal(qt, test)
# [1] TRUE